File: codes/leetcode/132Pattern.cpp
/*
************************************************
username    :   smmehrab
fullname    :   s.m.mehrabul islam
email       :   mehrab.24csedu.001@gmail.com
institute   :   university of dhaka, bangladesh
session     :   2017-2018
************************************************
*/
class Solution {
public:
    bool find132pattern(vector<int>& numbers) {
        // variables named according to 132
        
        int n = numbers.size();
        reverse(numbers.begin(), numbers.end());
        
        stack<int> twos;
        int two = INT_MIN;
        
        for(int one : numbers){
            if(one<two) return true;
            while(!twos.empty() && twos.top()<one) {
                two = twos.top();
                twos.pop();
            }
            twos.push(one);
        }
        return false;
    }
};

File: codes/leetcode/132Pattern.py
# ************************************************
# username  :   smmehrab
# fullname  :   s.m.mehrabul islam
# email     :   smmehrabul-2017614964@cs.du.ac.bd
# institute :   university of dhaka, bangladesh
# reg       :   2017614964
# ************************************************

from collections import deque

class Solution(object):
    def find132pattern(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        stack = deque()
        current_min = pow(10, 9) + 7

        for num in nums:
            # num is 3
            # finding 2
            while stack and num >= stack[-1][0]:
                stack.pop()
            # finding 1
            if stack and num > stack[-1][1]:
                return True
            # stacking 3 (possible 1 or 2 for the next 3s)
            stack.append((num, current_min))
            current_min = min(current_min, num)

        return False